pkg/operator/controller/configurable-route/controller.go

```diff
@@ -0,0 +1,393 @@
+	errors := []error{}
+
+	existingHashes := map[string]struct{}{}
+	for _, cr := range componentRoutes {
+		existingHashes[cr.Hash] = struct{}{}
+	}
+
+	roleBindingList := &rbacv1.RoleBindingList{}
+	r.cache.List(context.TODO(), roleBindingList, allComponentRouteResources()...)
+	for _, item := range roleBindingList.Items {
+		expectedHash, ok := item.GetLabels()[componentRouteHashAnnotation]
+		if !ok {
+			errors = append(errors, fmt.Errorf("Unable to find componentRoute hash label on roleBinding %s/%s", item.Namespace, item.Name))
+			continue
+		}
+
+		if _, ok := existingHashes[expectedHash]; !ok {
+			log.Info("deleting roleBinding", "name", item.GetName(), "namespace", item.GetNamespace())
+			if err := r.client.Delete(context.TODO(), &item); err != nil && !apierrors.IsNotFound(err) {
+				errors = append(errors, err)
+			}
+		}
+	}
+
+	return errors
+}
+
+func (r *reconciler) ensureServiceCertKeyPairSecretRole(owner metav1.Object, componentRoute aggregatedComponentRoute) (string, error) {
+	role := &rbacv1.Role{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: componentRoute.Name + "-",
+			Namespace:    r.config.SecretNamespace,
+			Labels: map[string]string{
+				componentRouteHashAnnotation: componentRoute.Hash,
+			},
+		},
+		Rules: []rbacv1.PolicyRule{
+			{
+				Verbs:         []string{"get", "list", "watch"},
+				APIGroups:     []string{""},
+				Resources:     []string{"secrets"},
+				ResourceNames: []string{componentRoute.ServingCertificateName},
+			},
+		},
+	}
+
+	roleList := &rbacv1.RoleList{}
+	if err := r.cache.List(context.TODO(), roleList, componentRouteResources(componentRoute)...); err != nil {
+		return "", err
+	}
+
+	if len(roleList.Items) == 0 {
+		if err := r.client.Create(context.TODO(), role); err != nil {
+			return "", err
+		}
+	} else {
+		role.Name = roleList.Items[0].Name
+		role.GenerateName = ""
+		if _, _, err := resourceapply.ApplyRole(r.kclient.RbacV1(), r.eventRecorder, role); err != nil {
+			return "", err
+		}
+	}
+
+	return role.GetName(), nil
+}
+
+func (r *reconciler) ensureServiceCertKeyPairSecretRoleBinding(owner metav1.Object, componentRoute aggregatedComponentRoute, roleName string) error {
+	roleBinding := &rbacv1.RoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      roleName,
+			Namespace: r.config.SecretNamespace,
+			Labels: map[string]string{
+				componentRouteHashAnnotation: componentRoute.Hash,
+			},
+		},
+		Subjects: componentRoute.GetSubjects(),
+		RoleRef: rbacv1.RoleRef{
+			Kind:     "Role",
+			Name:     roleName,
+			APIGroup: "rbac.authorization.k8s.io",
+		},
+	}
+
+	roleBindingList := &rbacv1.RoleBindingList{}
+	if err := r.cache.List(context.TODO(), roleBindingList, componentRouteResources(componentRoute)...); err != nil {
+		return err
+	}
+
+	if len(roleBindingList.Items) == 1 {
+		roleBinding.Name = roleBindingList.Items[0].Name
+	}
+
+	_, _, err := resourceapply.ApplyRoleBinding(r.kclient.RbacV1(), r.eventRecorder, roleBinding)
```

Comment:
This logic seems a little wonky.  Presumably, `roleBinding.Name` and `roleBindingList.Items[0].Name` could have differed before this assignment, or else you wouldn't have written it, right?  Would it be simpler always to use the name of the just-created role (so the role name/reference and rolebinding name would always match)?  

Is this logic's purpose to solve the following problem?

1. User creates a route configuration.
2. Controller creates role and rolebinding.
3. User deletes route configuration.
4. Controller deletes the role and fails to delete the rolebinding.
5. User creates the same route configuration again.
6. Controller creates a new role, and you want it to re-use the old rolebinding instead of creating a new one.

This highlights that the life cycles of these resources are potentially nuanced and confusing.  A comment would be helpful.  Maybe it would be better still to create the rolebinding with an owner reference to the role, and let the garbage collector sort it out; that would simplify this logic, and obviate the rolebinding cleanup logic as well.

Reply:
>Maybe it would be better still to create the rolebinding with an owner reference to the role, and let the garbage collector sort it out; that would simplify this logic, and obviate the rolebinding cleanup logic as well.

I agree with this approach and have implemented it in the latest version of the PR.